File: backtracking/egg.cpp
#include<bits/stdc++.h>
using namespace std;

int n, s[10], w[10], ans = 0;

void dfs(int k) {
    if(k == n) {
        int cnt = 0;
        for(int i = 0; i < n; i++) 
            if(s[i] <= 0) cnt++;
        ans = max(ans, cnt);
        return;
    }
    
    if(s[k] <= 0) {
        dfs(k + 1);
        return;
    }
    
    bool canHit = false;
    for(int i = 0; i < n; i++) {
        if(i != k && s[i] > 0) {
            canHit = true;
            s[k] -= w[i]; s[i] -= w[k];
            dfs(k + 1);
            s[k] += w[i]; s[i] += w[k];
        }
    }
    
    if(!canHit) {
        int cnt = 0;
        for(int i = 0; i < n; i++) 
            if(s[i] <= 0) cnt++;
        ans = max(ans, cnt);
    }
}

int main() {
    cin >> n;
    for(int i = 0; i < n; i++) cin >> s[i] >> w[i];
    dfs(0);
    cout << ans;
}

File: backtracking/egg.py
import sys
input = sys.stdin.readline
from collections import deque

ans = 0

def solve(k,n,s,w):
    global ans
    if k==n:
        cnt = 0
        for i in range(n):
            if s[i]<=0: cnt += 1
        ans = max(ans, cnt)
        return
    
    if s[k]<=0:
        solve(k+1,n,s,w)
        return

    canHit = False
    for i in range(n):
        if i!=k and s[i]>0:
            canHit = True
            s[k] -= w[i]
            s[i] -= w[k]
            solve(k+1,n,s,w)
            s[k] += w[i]
            s[i] += w[k]

    if not canHit:
        cnt = 0
        for i in range(n):
            if s[i]<=0: cnt+=1
        ans = max(ans,cnt)

def main():
    global ans
    ans = 0
    n = int(input())
    s = [0]*n
    w = [0]*n
    for i in range(n):
        s[i],w[i] = map(int, input().split())
    solve(0,n,s,w)
    print(ans)

if __name__ =="__main__":
    main()
